docs/en/concepts-and-designs/lal.md

```diff
@@ -0,0 +1,239 @@
+        // this is just a demo pattern
+    }
+    extractor {
+        tag key: "level", val: parsed.level
+        // we add a tag called `level` and its value is parsed.level, captured from the regexp above
+        traceId parsed.traceId
+        // we also extract the trace id from the parsed result, which will be used to associate the log with the trace
+    }
+    // ...
+}
+```
+
+- `grok`
+
+<!-- TODO: grok Java library has poor performance, need to benchmark it, the idea is basically the same with `regexp` above -->
+
+### Extractor
+
+Extractors aim to extract metadata from the logs. The metadata can be a service name, a service instance name, an
+endpoint name, or even a trace ID, all of which can be associated with the existing traces and metrics.
+
+- `service`
+
+`service` extracts the service name from the `parsed` result, and set it into the `LogData`, which will be persisted (if
+not dropped) and is used to associate with traces / metrics.
+
+- `instance`
+
+`instance` extracts the service instance name from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `endpoint`
+
+`endpoint` extracts the service instance name from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `traceId`
+
+`traceId` extracts the trace ID from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `segmentId`
+
+`segmentId` extracts the segment ID from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `spanId`
+
+`spanId` extracts the span ID from the `parsed` result, and set it into the `LogData`, which will be
+persisted (if not dropped) and is used to associate with traces / metrics.
+
+- `metrics`
+
+`metrics` extracts / generates metrics from the logs, and sends the generated metrics to the meter system, you can configure [MAL](mal.md) for further analysis of these metrics. Examples are as follows:
+
+```groovy
+filter {
+    // ...
+    extractor {
+        service parsed.serviceName
+        metrics("log_count") {
+            timestamp: parsed.timestamp
+            tags: ["level": parsed.level, "service": parsed.service, "instance": parsed.instance]
+            value: 1
+        }
+        metrics("http_response_time") {
+            timestamp: parsed.timestamp
+            tags: ["status_code": parsed.statusCode, "service": parsed.service, "instance": parsed.instance]
+            value: parsed.duration
+        }
+    }
+    // ...
+}
+```
+
+The extractor above generates a metrics named `log_count`, with tag key `level` and value `1`, after this, you can configure MAL rules to calculate the log count grouping by logging level like this:
+
+```yaml
```

Comment:
Where should user activate the MAL script? Is this in the log-analyzer? Or some places else?

Reply:
MAL was related to receivers, you could see the rule YAMLs existing in different folders, which are controlled by different receivers.

Reply:
> Where should user activate the MAL script? Is this in the log-analyzer? Or some places else?

I think it's ok to put them in log-analyzer

Reply:
Let's make this clear in the documentation. Should this MAL script file be pointed in the LAL?